### ISSUE TITLE
Improve performance of static code reusage

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#19318]
### Summary of the new feature / enhancement

The section [Avoid repeated calls to a function](https://learn.microsoft.com/powershell/scripting/dev-cross-plat/performance/script-authoring-considerations?view=powershell-7.3#avoid-repeated-calls-to-a-function) hits a good point. But there are quiet some situations were it is needed to (re)use the specific code block at several locations in a script, which leaves a dilemma:

Should I create a quiet expensive function for a easier manageable [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) code, or
Should I go for performance and copy the concerned piece of code block at several locations?

For details see stackoverflow question/answer: [What is the best way to reuse static code](https://stackoverflow.com/questions/75722846/what-is-the-best-way-to-reuse-static-code)

### Proposed technical implementation details (optional)

It would be good to be able to reuse specific code blocks at (almost) no extra costs

I am imaging a specific syntax to refer to an existing code block but it could also be a smarter dot-sourcing: in the case `CodeBlock` is assigned to a variable and the variable is dot-sourced (and there are no arguments in play), replace the variable with the literal code:

In other words:
```PowerShell
$MyCode = { "Some static code $Value" }
for ($Value = 0; $Value -lt 10; $Value++) {
    . $MyCode
}
```
Compile as:
```PowerShell
for ($Value = 0; $Value -lt 10; $Value++) {
    "Some static code $Value"
}
```

**Comment by 237dmitry** [comment]
> Should I go for performance and copy the concerned piece of code block at several locations?

The difference in speed is not very significant between the literal code and the invocation. The sourced script-block lags behind.
```powershell
 $ h -Count 3 | fl Commandline, Duration         

CommandLine : & {
              $out = for ($Value = 0; $Value -lt 10000; $Value++) { "Some static code $Value" }; $out
              }
Duration    : 00:00:00.2244007

CommandLine : & {
              $MyCode = { "Some static code $Value" }
              $out = for ($Value = 0; $Value -lt 10000; $Value++) { $MyCode.Invoke() }; $out
              }          
Duration    : 00:00:00.2992577

CommandLine : & {
              $MyCode = { "Some static code $Value" }
              $out = for ($Value = 0; $Value -lt 10000; $Value++) { . $MyCode }; $out
              }
Duration    : 00:00:00.4969723
```

**Comment by SeeminglyScience** [comment]
I would caution against microbenchmarking PowerShell in general. There are a ton of pitfalls that can skew results or in general be misleading compared to real world code.

A static method will likely be the fastest (other than including inline) for most work loads. I'd personally suggest using a type nested in the function if the code is not used outside of it.

For the proposal, I'd recommend finding some notable examples of existing code that could benefit from something like this. The compiler is one of the most complicated parts of the code base and this change would require a decent chunk of time from someone who knows it very well. The bar for impact will be pretty high

**Comment by jhoneill** [comment]
@SeeminglyScience  put it well 

This structure
```
for ($Value = 0; $Value -lt $x ; $Value++) {
    . $mycode
}
```
Is almost always bad, because (a) there are a million call/returns  (b) we have no idea what is happening in the loop. If it is a function which acts as a "black box" where we don't need to know how something is implemented AND it is being called a small number of times then it is OK. 
But if we need to go and find $myCode to understand what the whole does. (i.e. it isn't a black box) then it's just spaghetti code.  A goto by other means. (well a goSub, really).   Sadly some people learn practises which make their code jump about when it doesn't need need to and is just harder to understand as a result

````
for ($i = 0; $i -lt $x ; $i++) {
    transferMoney $National_transactions[$i]
}
```
Might be acceptable especially if $x is always small, and if we don't need to know _how_ a transfer is done, and is _preferable_ if transfer is called from multiple places.  But writing a function which we expect to process many items one after the other, but where each must be passed singly would be an mistake. `transferMoney` in my example should be written to accept an array so it can be called without the loop. 
```
    transferMoney $National_transactions[0..$x]
   ... 
    transferMoney $International_transactions[0..$y]
```
